vm/actor/src/builtin/puppet/mod.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use num_derive::FromPrimitive;
+use num_traits::FromPrimitive;
+
+use crate::check_empty_params;
+use address::Address;
+use encoding::{tuple::*, Cbor};
+use ipld_blockstore::BlockStore;
+use num_bigint::bigint_ser;
+use runtime::{ActorCode, Runtime};
+use serde::{Deserialize, Serialize};
+use vm::{ActorError, ExitCode, MethodNum, Serialized, TokenAmount, METHOD_CONSTRUCTOR};
+
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum Method {
+    Constructor = METHOD_CONSTRUCTOR,
+    SendMethod = 2,
+    SendMarshalCBORFailure = 3,
+    ReturnMarshalCBORFailure = 4,
+    RuntimeTransactionMarshalCBORFailure = 5,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SendParams {
+    pub to: Address,
+    #[serde(with = "bigint_ser")]
+    pub value: TokenAmount,
+    pub method: u64,
+    pub params: Vec<u8>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SendReturn {
+    pub return_bytes: Serialized,
+    pub code: ExitCode,
+}
+
+#[derive(Default, Serialize, Deserialize)]
+pub struct FailToMarshalCBOR {}
+
+impl Cbor for FailToMarshalCBOR {}
+
+#[derive(Default, Serialize_tuple, Deserialize_tuple)]
+pub struct State {
+    opt_fail: Vec<Option<FailToMarshalCBOR>>,
+}
+
+impl Cbor for State {}
+
+pub struct Actor;
+
+impl Actor {
+    fn constructor<BS, RT>(rt: &mut RT) -> Result<(), ActorError>
+    where
+        BS: BlockStore,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_accept_any()?;
+
+        // TODO revisit based on issue: https://github.com/filecoin-project/specs-actors/issues/317
```

Comment:
what does this have to do with it?

Reply:
my bad copied over the comment by accident